src/consensus/cfund.cpp

```diff
@@ -221,28 +221,95 @@ bool CFund::IsValidPaymentRequest(CTransaction tx, CCoinsViewCache& coins, int n
 
 }
 
+flags CFund::CPaymentRequest::GetLastState() const {
```

Comment:
As i understand it we're storing the block height of each state change into a <blockhash,state> map against the payment request (and proposal) now correct? How was this done previously? Was it not needed? Did we just kept the last state and the last blockhash?

Reply:
yes, we store a map of block hash and states (the whole state history, so its easy to roll back).
before we just stored the last state and blockhash, which was the main source of issues related to reorganizations (how to recover previous state)